Doc/library/typing.rst

```diff
@@ -932,6 +932,24 @@ The module defines the following classes, functions and decorators:
    This wraps the decorator with something that wraps the decorated
    function in :func:`no_type_check`.
 
+.. decorator:: type_check_only
+
+   Decorator to mark a class or function to be unavailable at runtime.
+
+   This decorator is itself not available at runtime. It is mainly
+   intended to mark classes that are defined in type stub files if
+   an implementation returns an instance of a private class::
+
+      @type_check_only
+      class Response:  # private or not available at runtime
+          code: int
+          def get_header(self, name: str) -> str: ...
+
+      def fetch_response() -> Response: ...
+
+    Note that returning instances of private classes is not recommended.
```

Comment:
Even if we decide that documenting this makes sense, providing an example that is not recommended is a bad idea. I would instead provide a more realistic example. IIUC a typical situation for this would be a private protocol for argument types. Protocols are not accepted yet, so maybe this should wait until they are?

Reply:
In my experience, returning instances of private classes is a pretty common occurence. There are multiple examples of it in the standard library, the IO classes as well as `Pattern` and `Match` being the most prominent.